### PR TITLE
feat: Add e2e reorg tests to syncers 

### PR DESCRIPTION
**Pull request by rbpol**
PR adds new `e2e` tests for `l1infotreesyncer` that test the syncer behavior with reorgs.

Also fixes a couple of bugs:
- deadlocks in `reorgdetector`.
- missing reorg of a `l1info_leaf` table in the `l1infotreesyncer`.
- bugs in regards to context being canceled while the downloader is pooling for blocks, which caused a fatal after max number of attempts to pull some data failed, even though the context was canceled.
- removed unnecessary emptying of the download channel in the `handleReorg` function in the syncer.
- rebuilding the tree had an issue on `initCache` function, which is called when reorg happens.

### CODE REVIEW

File: l1infotreesync/e2e_test.go [hunk 1]
```diff
@@ -182,58 +292,50 @@ func TestStressAndReorgs(t *testing.T) {
...
+
+		// Update L1 Info Tree + Rollup Exit Tree
+		newLocalExitRoot := common.HexToHash(strconv.Itoa(i) + "ffff" + strconv.Itoa(j))
+		_, err = verifySC.VerifyBatches(auth, rollupID, 0, newLocalExitRoot, common.Hash{}, true)
+		require.NoError(t, err)
+
+		// Update Rollup Exit Tree
+		newLocalExitRoot = common.HexToHash(strconv.Itoa(i) + "fffa" + strconv.Itoa(j))
+		_, err = verifySC.VerifyBatches(auth, rollupID, 0, newLocalExitRoot, common.Hash{}, false)
+		require.NoError(t, err)
```
Comment (Stefan-Ethernal) [inline l1infotreesync/e2e_test.go:312]: Shouldn't you check the `i` here and based on modulus (0, 1, 2) perform either `Update L1 Info Tree`, `Update L1 Info Tree + Rollup Exit Tree` or `Update Rollup Exit Tree` (as it was before)?

File: tree/tree_test.go [hunk 2]
```diff
@@ -18,6 +19,88 @@ import (
...
+		require.NoError(t, err)
+		require.Equal(t, expectedRoot.Hash, root2.Hash)
+		require.Equal(t, expectedRoot.Index, root2.Index)
+	})
+
+	t.Run("Check after rebuild tree when reorg", func(t *testing.T) {
+		numOfLeavesToAdd := 10
+		indexToCheck := uint32(numOfLeavesToAdd - 1)
+		treeDB := createTreeDB()
+		merkletree := tree.NewAppendOnlyTree(treeDB, "")
```
Comment (Stefan-Ethernal) [inline tree/tree_test.go:73]: Nit
```suggestion
		merkleTree := tree.NewAppendOnlyTree(treeDB, "")
```

File: sync/evmdriver.go [hunk 3]
```diff
@@ -84,18 +87,20 @@ reset:
...
 		select {
 		case b := <-downloadCh:
-			d.log.Debug("handleNewBlock")
+			d.log.Debug("handleNewBlock: ", b.Num, b.Hash, " syncID ", syncID)
 			d.handleNewBlock(ctx, b)
+			d.log.Debug("handleNewBlock done: ", b.Num, b.Hash, " syncID ", syncID)
 		case firstReorgedBlock := <-d.reorgSub.ReorgedBlock:
-			d.log.Debug("handleReorg")
-			d.handleReorg(ctx, cancel, downloadCh, firstReorgedBlock)
+			d.log.Debug("handleReorg: ", firstReorgedBlock, " syncID ", syncID)
```
Comment (Stefan-Ethernal) [inline sync/evmdriver.go:102]: Correct me if I'm wrong, but shouldn't there be an odd number of parameters?

```suggestion
			d.log.Debug("handleReorg: ", "first reorged block", firstReorgedBlock, "syncID", syncID)
```

File: sync/evmdriver.go [hunk 4]
```diff
@@ -84,18 +87,20 @@ reset:
...
 	// start downloading
 	downloadCh := make(chan EVMBlock, d.downloadBufferSize)
-	go d.downloader.Download(cancellableCtx, lastProcessedBlock, downloadCh)
+	go d.downloader.Download(cancellableCtx, lastProcessedBlock+1, downloadCh)
 
 	for {
 		select {
 		case b := <-downloadCh:
-			d.log.Debug("handleNewBlock")
+			d.log.Debug("handleNewBlock: ", b.Num, b.Hash, " syncID ", syncID)
```
Comment (Stefan-Ethernal) [inline sync/evmdriver.go:98]: Double check, but should be alright
```suggestion
			d.log.Debug("handleNewBlock:", b.Num, b.Hash, "syncID", syncID)
```

File: sync/evmdriver.go [hunk 5]
```diff
@@ -84,18 +87,20 @@ reset:
 	cancellableCtx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
+	log.Info("Starting sync. syncID: ", syncID, " lastProcessedBlock", lastProcessedBlock)
```
Comment (Stefan-Ethernal) [inline sync/evmdriver.go:90]: Double check
```suggestion
	log.Info("Starting sync...", "syncID: ", syncID, "lastProcessedBlock", lastProcessedBlock)
```

File: reorgdetector/reorgdetector.go [hunk 6]
```diff
@@ -153,6 +161,8 @@ func (rd *ReorgDetector) detectReorgInTrackedList(ctx context.Context) error {
 					continue
 				}
 
+				log.Info("[ReorgDetector] Reorg detected", " blockNum ", hdr.Num)
```
Comment (Stefan-Ethernal) [inline reorgdetector/reorgdetector.go:164]:
```suggestion
				log.Info("[ReorgDetector] Reorg detected", "blockNum", hdr.Num)
```

File: l1infotreesync/e2e_test.go (hunk 1 above)
Comment (goran-ethernal) [reply l1infotreesync/e2e_test.go:312]: It doesn't matter, it is just a inserting test events. I would leave it like this.

File: sync/evmdriver.go (hunk 3 above)
Comment (goran-ethernal) [reply sync/evmdriver.go:102]: No, it is arbitrary.

File: reorgdetector/reorgdetector.go (hunk 6 above)
Comment (goran-ethernal) [reply reorgdetector/reorgdetector.go:164]: This log was also added for debugging. I will remove it.

File: l1infotreesync/processor.go [hunk 7]
```diff
@@ -211,6 +211,11 @@ func (p *processor) Reorg(ctx context.Context, firstReorgedBlock uint64) error {
 		return err
 	}
 
+	_, err = tx.Exec(`DELETE FROM l1info_leaf WHERE block_num >= $1;`, firstReorgedBlock)
```
[the remark this answers is not in the document]
Comment (goran-ethernal) [reply l1infotreesync/processor.go]: Fix: https://github.com/0xPolygon/cdk/pull/56/commits/b754a8b397e29621456a9af15aad0e55984b02cd

File: l1infotreesync/downloader.go [hunk 8]
```diff
@@ -86,7 +86,8 @@ func buildAppender(client EthClienter, globalExitRoot, rollupManager common.Addr
 				l, err,
 			)
 		}
-		log.Infof("updateL1InfoTreeSignatureV2: expected root: %s", common.Bytes2Hex(l1InfoTreeUpdate.CurrentL1InfoRoot[:]))
+		log.Infof("updateL1InfoTreeSignatureV2: expected root: %s",
+			common.BytesToHash(l1InfoTreeUpdate.CurrentL1InfoRoot[:]).String())
```
[the remark this answers is not in the document]
Comment (goran-ethernal) [reply l1infotreesync/downloader.go]: Fix: https://github.com/0xPolygon/cdk/pull/56/commits/b754a8b397e29621456a9af15aad0e55984b02cd

File: sync/evmdriver.go (hunk 4 above)
Comment (goran-ethernal) [reply sync/evmdriver.go]: Fix: https://github.com/0xPolygon/cdk/pull/56/commits/b754a8b397e29621456a9af15aad0e55984b02cd

File: sync/evmdriver.go (hunk 5 above)
Comment (goran-ethernal) [reply sync/evmdriver.go]: Fix: https://github.com/0xPolygon/cdk/pull/56/commits/b754a8b397e29621456a9af15aad0e55984b02cd

File: reorgdetector/reorgdetector.go (hunk 6 above)
Comment (goran-ethernal) [reply reorgdetector/reorgdetector.go]: Fix: https://github.com/0xPolygon/cdk/pull/56/commits/b754a8b397e29621456a9af15aad0e55984b02cd

**Pull request merged**